cuda/test/base/math.cu

```diff
@@ -0,0 +1,139 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2019, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include <cmath>
+#include <complex>
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+
+
+#include "cuda/base/math.hpp"
+
+
+namespace {
+
+
+template <typename T>
+__global__ void test_real_isfinite(bool *result)
+{
+    constexpr auto inf = INFINITY;
+    bool test_true{};
+    bool test_false{};
+
+    test_true =
+        gko::isfinite(T{0}) && gko::isfinite(-T{0}) && gko::isfinite(T{1});
+    test_false =
+        gko::isfinite(inf) || gko::isfinite(-inf) || gko::isfinite(NAN);
+    *result = test_true && !test_false;
+}
+
+
+template <typename T>
+__global__ void test_complex_isfinite(bool *result)
+{
+    using c_type = thrust::complex<T>;  // std::complex<T>;
+    constexpr T inf = INFINITY;
+    constexpr T quiet_nan = NAN;
+    bool test_true{};
+    bool test_false{};
+
+    test_true = gko::isfinite(c_type{T{0}, T{0}}) &&
+                gko::isfinite(c_type{-T{0}, -T{0}}) &&
+                gko::isfinite(c_type{T{1}, T{0}}) &&
+                gko::isfinite(c_type{T{0}, T{1}});
+    test_false =
+        gko::isfinite(c_type{inf, T{0}}) || gko::isfinite(c_type{-inf, T{0}}) ||
```

Comment:
Since this is inside a kernel, it is not easy to test everything separately with the GTest integration (I would probably need to add one parameter per test).
I thought this way, it is easier to read the test (and not make mistakes while writing it). Also, if there is a problem with it, you could just comment out parts and see where the issue is (I know that this is a bit annoying, but this test should not fail randomly after some time).

But if you want, I can do one parameter for each assert, or do 2 parameters in total, one for detecting non-finite, and one for detecting actual finite numbers.

Reply:
No, this looks fine.